Tidy-up the testsuite/manual commands

dra27 opened the pull request:
A small piece of tidying spun out from various other bits of build-system-related work.

`DIFF` is no longer used anywhere, so remove the Windows-specific command (`diff` is invoked directly by `ocamltest`).

Move the default for `FIND` from `testsuite/Makefile` to `Makefile.config.in` so that both the Windows and Unix variants are in the same file

`SET_LD_PATH` is only used by the manual build system; `SORT`, `DIFF` and `CYGPATH` are only used by the testsuite so these definitions could finally move to their respective Makefiles?

dra27 committed:
Tidy-up the testsuite/manual commands

DIFF is no longer used anywhere, so remove the Windows-specific command.
Move the default for FIND from testsuite/Makefile to Makefile.config.in.

SET_LD_PATH is now only used by the manual build system.

dra27 committed:
Remove $(FIND) completely

It was only used in a clean target, where it wasn't read anyway!

dra27 committed:
Move $(CYGPATH) and $(SORT) to testsuite/Makefile

dra27 committed:
Move $(SET_LD_PATH) to Makefile.common

dra27 commented:
Edited - `$(FIND)` actually didn't work (it was ignored), so I've removed it completely. Having done that, I've also moved `$(SORT)` and `$(CYGPATH)` to `testsuite/Makefile` as that's the only place they're used.

I've then moved `$(SET_LD_PATH)` to `Makefile.common`. It should be possible to get rid of it completely, but that's for another day.

shindere commented:
Yet another nice one, thanks!

Perhaps sme of the commits oculd be squashed?

Also, two questions:

 1. How much work would it be to remove `SET_LD_PATH` completely? Would happily help, here.
 2. Could cygpath and sort be checked by configure? I thinkk that's what I was suggested when I wnated to touch `diff`.

dra27 commented:
Thanks, @shindere! The four commits were intended to make reviewing slightly easier, as the overall diff is just a big deletion in `Makefile.config.in` - they certainly can be squashed on merge.

I don't know what needs to be done to eliminate `SET_LD_PATH` in the manual's build system, to be honest! It's no longer needed in the main build system because `ocamltest` sets `CAML_LD_LIBRARY_PATH` so I expect it would continue to be needed in the manual as long as it has need of binaries using `otherlibs`.

Up to you for `SORT` and `CYGPATH`. I would probably err on keeping it in `testsuite/Makefile` - in this case I don't think we're checking for these commands, they're required. The `DIFF` case is different (I've got a TODO item tracking that for "some point"!) because it would be good to probe what the diff command actually is and what flags it supports (for example, if a test fails on Alpine at the moment, we don't get the diff at all because we pass an invalid flag to it).

shindere commented:
David Allsopp (2021/07/20 08:47 -0700):
> Thanks, @shindere! The four commits were intended to make reviewing
> slightly easier, as the overall diff is just a big deletion in
> `Makefile.config.in` - they certainly can be squashed on merge.

I'd do that but feel freeto not do it. ;-)

> I don't know what needs to be done to eliminate `SET_LD_PATH` in the
> manual's build system, to be honest! It's no longer needed in the main
> build system because `ocamltest` sets `CAML_LD_LIBRARY_PATH` so I
> expect it would continue to be needed in the manual as long as it has
> need of binaries using `otherlibs`.

Would you mind adding a comment somewhere in the manual's makefile
saying that this remains to be done. Do we have a convention all of us
agree on for TODOs in the compiler's codebase, by the way?

> Up to you for `SORT` and `CYGPATH`. I would probably err on keeping it
> in `testsuite/Makefile` - in this case I don't think we're checking
> for these commands, they're required.

Or perhaps we should check for them if ocamltestis enabled? Or perhaps
we could introduce an `-enable-testsuite` option that would check for
the presence of these commands andfail if they are not found, since they
are required to be able to run the testsuite, as I understand it?

> The `DIFF` case is different
> (I've got a TODO item tracking that for "some point"!) because it
> would be good to probe what the diff command actually is and what
> flags it supports (for example, if a test fails on Alpine at the
> moment, we don't get the diff at all because we pass an invalid flag
> to it).

I see. This kind of brings me back to the question above about how we
best could share our TODO items. It's a pity that this one is on your
list whereas maybe somebody would feel like dealing with it. Wellthere
may be a PR or an issue opened about that already but I do'nt knwo
whether such things are best shared as issues or with comments in the
codebase? I just think it would be niceto make them public.

NOt to say any of those shsould go in this PR. Except if you agree about
the SET_LD_LIBRARY thing. If you prefer, you could also create an issue
and assign it to me and leave thepresentPR as it is.

shindere commented:
Just approved. Feel free to merge, not matter how you want to deal with
SET_LD_LIBRARY_PATH thing.

dra27 committed:
Remove unnecessary SET_LD_PATH calls in manual

CAML_LD_LIBRARY_PATH only needs to be messed around with for utilities
which use otherlibs.

dra27 committed:
Eliminate SET_LD_PATH completely

Use ocamlrun -I instead (which has higher precedence than
CAML_LD_LIBRARY_PATH).

dra27 commented:
Actually, rather than finding a place to put a "TODO" item for the manual, some other work I've been doing meant I realised the easy fix! Two extra commits pushed - the first one removes some unnecessary `SET_LD_PATH` calls (the programs being invoked do not use otherlibs, so don't need it) and then another commit which converts `CAML_LD_LIBRARY_PATH=$(ROOTDIR)/otherlibs/unix:$(ROOTDIR)/otherlibs/str $(ROOTDIR)/boot/ocamlrun` into `$(ROOTDIR)/boot/ocamlrun -I $(ROOTDIR)/otherlibs/unix -I $(ROOTDIR)/otherlibs/str` which is a good deal simpler. There's no need to worry then at all about `CAML_LD_LIBRARY_PATH` since `-I` directories are searched before `CAML_LD_LIBRARY_PATH` ones.

dra27 commented:
I don't know about TODO lists in general - I've no problem with sharing them. It's worth saying that are gazillions of small things like that on my "TODO" list which won't necessarily ever get done, so it never bothers me if someone else gets there first.

dra27 commented:
Where `DIFF` is concerned, I expect that the next time a test fails on Alpine only and I can't see why because the `diff` command is working then I'll probably fix it 🙂

shindere commented:
Awesome David! Thanks!

Regarding the way to store TODOs, at first I thought that issues mith a
dedicated label would be a nice way to do that but then I realised that
if such issues point specific places in the code and the code is then
modified, then the location mentionned by the issues will become wrong
which could be annoying. So I think, if we can agree on a way to share
TODOs that affect speecific parts of the code, then those should be
stored in the code itself.

shindere commented:
David Allsopp (2021/08/11 01:20 -0700):
> I don't know about TODO lists in general - I've no problem with
> sharing them. It's worth saying that are gazillions of small things
> like that on my "TODO" list which won't necessarily ever get done, so
> it never bothers me if someone else gets there first.

I'd like to share TODOs! :) I don't know how but I would like to!

shindere commented:
David Allsopp (2021/08/11 01:21 -0700):
> Where `DIFF` is concerned, I expect that the next time a test fails on
> Alpine only and I can't see why because the `diff` command is working
> then I'll probably fix it 🙂

OK. Thanks.

dra27 commented:
Actually, the `DIFF` issue is a bug, so I have opened https://github.com/ocaml/ocaml/issues/10562 to track that!

dra27 commented:
I'm just double-checking that the manual builds correctly (the main CI check is disabled at the moment), but are you otherwise happy for this to be squashed and merged?

shindere commented:
David Allsopp (2021/08/11 04:16 -0700):
> I'm just double-checking that the manual builds correctly (the main CI
> check is disabled at the moment), but are you otherwise happy for this
> to squashed and merged?

Yes, definitely! Very happy, actually. Big thanks!

dra27 commented:
Excellent, the manual build is working fine in [this run](https://github.com/dra27/ocaml/runs/3300405410)
